File: connectors/connector-basecamp/src/main/scala/com/taskadapter/connector/basecamp/BasecampToGTask.scala
package com.taskadapter.connector.basecamp

import com.taskadapter.connector.definition.TaskId
import com.taskadapter.model._
import org.json.JSONObject

object BasecampToGTask {
  def parseTask(obj: JSONObject): GTask = {
    val result = new GTask
    val id = JsonUtils.getInt("id", obj)
    result.setId(id.toLong)
    result.setKey(id.toString)
    // must set source system id, otherwise "update task" is impossible later
    result.setSourceSystemId(new TaskId(id.toLong, id.toString))

    result.setValue(BasecampField.content, JsonUtils.getOptString("content", obj))
    val completedFloatValue: java.lang.Float = if (JsonUtils.getOptBool("completed", obj)) 100f else 0f
    result.setValue(AllFields.doneRatio, completedFloatValue)
    result.setValue(AllFields.dueDate, JsonUtils.getOptShortDate("due_at", obj))
    result.setValue(AllFields.createdOn, JsonUtils.getOptLongDate("created_at", obj))
    result.setValue(AllFields.updatedOn, JsonUtils.getOptLongDate("updated_at", obj))
    val assObj = JsonUtils.getOptObject("assignee", obj)
    if (assObj != null) {
      result.setValue(AllFields.assigneeFullName, parseUser(assObj).getDisplayName)
    }
    result
  }

  def parseUser(assObj: JSONObject): GUser = {
    new GUser().setId(JsonUtils.getInt("id", assObj))
      .setDisplayName(JsonUtils.getOptString("name", assObj))
  }
}

File: connectors/connector-basecamp/src/main/java/com/taskadapter/connector/basecamp/BasecampToGTask.java
package com.taskadapter.connector.basecamp;

import com.taskadapter.connector.definition.TaskId;
import com.taskadapter.connector.definition.exceptions.CommunicationException;
import com.taskadapter.model.AllFields;
import com.taskadapter.model.GTask;
import com.taskadapter.model.GUser;
import org.json.JSONObject;

public class BasecampToGTask {
    public static GTask parseTask(JSONObject obj) throws CommunicationException {
        var result = new GTask();
        var id = JsonUtils.getInt("id", obj);
        long longId = id;
        result.setId(longId);
        String stringId = id + "";
        result.setKey(stringId);
        // must set source system id, otherwise "update task" is impossible later
        result.setSourceSystemId(new TaskId(longId, stringId));

        result.setValue(BasecampField.content, JsonUtils.getOptString("content", obj));
        Float completedFloatValue = JsonUtils.getOptBool("completed", obj) ? 100f : 0f;
        result.setValue(AllFields.doneRatio, completedFloatValue);
        result.setValue(AllFields.dueDate, JsonUtils.getOptShortDate("due_at", obj));
        result.setValue(AllFields.createdOn, JsonUtils.getOptLongDate("created_at", obj));
        result.setValue(AllFields.updatedOn, JsonUtils.getOptLongDate("updated_at", obj));
        var assObj = JsonUtils.getOptObject("assignee", obj);
        if (assObj != null) {
            result.setValue(AllFields.assigneeFullName, parseUser(assObj).getDisplayName());
        }
        return result;
    }

    static GUser parseUser(JSONObject assObj) throws CommunicationException {
        return new GUser().setId(JsonUtils.getInt("id", assObj))
                .setDisplayName(JsonUtils.getOptString("name", assObj));
    }
}
